guides/https.md

```diff
@@ -0,0 +1,255 @@
+```
+
+Note that reading environment variables in Mix config files only works when starting the application using Mix, e.g. in a development environment. In production, a different approach is needed for runtime configuration, but this is out of scope for the current document.
+
+The certificate and CA chain can also be specified using DER binaries, using the `:cert` and `:cacerts` options, but this is best avoided. The use of PEM files has been tested much more thoroughly with the Erlang/OTP `:ssl` application, and there have been a number of issues with DER binary certificates in the past.
+
+## Custom Diffie-Hellman Parameters
+
+It is recommended to generate a custom set of Diffie-Hellman parameters, to be used for the DHE key exchange. Use the following OpenSSL CLI command to create a `dhparam.pem` file:
+
+`openssl dhparam -out dhparams.pem 4096`
+
+On a slow machine (e.g. a cheap VPS) this may take several hours. You may want to run the command on a strong machine and copy the file over to the target server: the file does not need to be kept secret. It is best practice to rotate the file periodically.
+
+Pass the (relative or absolute) path using the `:dhfile` option:
+
+```elixir
+Plug.Cowboy.https MyApp.MyPlug, [],
+  port: 8443,
+  cipher_suite: :strong,
+  certfile: "priv/cert/selfsigned.pem",
+  keyfile: "priv/cert/selfsigned_key.pem",
+  dhfile: "priv/cert/dhparams.pem",
+  otp_app: :my_app
+```
+
+If no custom parameters are specified, Erlang's `:ssl` uses its built-in defaults. Since OTP 19 this has been the 2048-bit 'group 14' from RFC3526.
+
+## Renewing Certificates
+
+Whenever a certificate is about to expire, when the contents of the certificate has been updated, or when the certificate is 're-keyed', the HTTPS server needs to be updated with the new certificate and/or key.
+
+When using the `:certfile` and `:keyfile` parameters to reference PEM files on disk, replacing the certificate and key is as simple as overwriting the files. Erlang's `:ssl` application periodically checks the timestamps of such files, and reloads them if it detects a change. It may be best to use symbolic links that point to versioned copies of the files, to allow for quick rollback in case of problems.
+
+Note that there is a potential race condition when both the certificate and the key need to be replaced at the same time: if the `:ssl` application detects the change of one file before the other file is updated, the partial update can leave the HTTPS server with a mismatched private key. This can be avoiding by placing the private key in the same PEM file as the certificate, and omitting the `:keyfile` option. This configuration allows atomic updates, and it works because `:ssl` looks for a private key entry in the `:certfile` PEM file if no `:key` or `:keyfile` option is specified.
+
+Also note that some filesystems, such as network and container filesystems or VM-mounted volumes, may not support reliable detection of file changes through metadata.
```

Comment:
Not sure this is true: I've seen such issues with filesystem notifiers, e.g. for Phoenix live reload, but the polling done by `:ssl` on the file attributes should work on any filesystem, no?

Reply:
Ok, so let's keep it for now